utils/src/bitmap/mod.rs

```diff
@@ -0,0 +1,1705 @@
+                        // For the last byte, only count bits up to next_bit % 8
+                        let bits_in_last_byte = self.next_bit % 8;
+                        let mask = (1u8 << bits_in_last_byte) - 1;
+                        count += (byte & mask).count_ones() as u64;
+                    } else {
+                        count += byte.count_ones() as u64;
+                    }
+                }
+            } else {
+                // For all other chunks, count all bits
+                for byte in chunk {
+                    count += byte.count_ones() as u64;
+                }
+            }
+        }
+        count
+    }
+
+    /// Returns the number of bits set to 0.
+    #[inline]
+    pub fn count_zeros(&self) -> u64 {
+        self.len() - self.count_ones()
+    }
+
+    /// Performs a bitwise AND with another BitMap.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the lengths don't match.
+    pub fn and(&mut self, other: &BitMap<N>) {
+        self.binary_op(other, |a, b| a & b);
+    }
+
+    /// Performs a bitwise OR with another BitMap.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the lengths don't match.
+    pub fn or(&mut self, other: &BitMap<N>) {
+        self.binary_op(other, |a, b| a | b);
+    }
+
+    /// Performs a bitwise XOR with another BitMap.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the lengths don't match.
+    pub fn xor(&mut self, other: &BitMap<N>) {
+        self.binary_op(other, |a, b| a ^ b);
+    }
+
+    /// Flips all bits (1s become 0s and vice versa).
+    pub fn invert(&mut self) {
+        for chunk in &mut self.chunks {
+            for byte in chunk {
+                *byte = !*byte;
+            }
+        }
+    }
+
+    /// Asserts that the index is within bounds.
+    #[inline(always)]
+    fn assert_index(&self, index: u64) {
+        assert!(index < self.len(), "Index out of bounds");
+    }
+
+    /// Asserts that the lengths of two [BitMap]s match.
+    #[inline(always)]
+    fn assert_eq_len(&self, other: &BitMap<N>) {
+        assert_eq!(self.len(), other.len(), "BitMap lengths don't match");
+    }
+
+    /// Creates an iterator over the bits.
+    pub fn iter(&self) -> Iterator<'_, N> {
+        Iterator {
+            bitmap: self,
+            pos: 0,
+        }
+    }
+
+    /// Helper for binary operations (AND, OR, XOR)
+    #[inline]
+    fn binary_op<F: Fn(u8, u8) -> u8>(&mut self, other: &BitMap<N>, op: F) {
+        self.assert_eq_len(other);
+        for (a_chunk, b_chunk) in self.chunks.iter_mut().zip(other.chunks.iter()) {
+            for (a_byte, b_byte) in a_chunk.iter_mut().zip(b_chunk.iter()) {
+                *a_byte = op(*a_byte, *b_byte);
+            }
+        }
+    }
+}
+
+impl<const N: usize> Default for BitMap<N> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<const N: usize> PartialEq for BitMap<N> {
+    fn eq(&self, other: &Self) -> bool {
```

Comment:
Does `Self` imply that `N` is the same between both? If so, perhaps just compare the chunks rather than the bits (for efficiency)

Keep in mind that it seems that bits beyond `next_bit` in the last chunk may not be equal between two "equal" bitmaps